docs/en/man/fms.1
.\" This file is part of FMS, an agent-based Financial Market Simulator
.\" FMS is (c) 2008,2009 Jean-Charles Bagneris. See LICENSE for redistribution
.\" information and usual disclaimer.
.\" Process with groff -man -Tascii fms.1
.\"
.TH FMS 1 "2008-12-08" "FMS Development Team" "Financial Market Simulator"
.SH NAME
fms \- a financial market simulator
.SH SYNOPSIS
.B startfms.py [options] [command] [
.I experiment-config-file
.B ]
.SH DESCRIPTION
.B fms
is an agent-based market simulator. The main program,
.B  fms.py
can read and validate an "experiment" config file, and/or
run it, outputting transactions data on 
.I stdout
or in a .csv file.
The experiment configuration file is described in experiment-conf.yml(5).
.SH OPTIONS
Options override
.I experiment-config-file
parameters values.
.IP "-h, --help"
Show help message and quit.
.IP "-L LEVEL, --loglevel LEVEL"
Set logging level to
.I LEVEL.
This overrides the -v option.
.IP "-v, --verbose"
Set logging level to 'info'.
Overrided by -L or --loglevel option.
.IP "--show-books, --show-limits"
Show best limits on each step.
.IP "--timer"
Show a day:time timer while experiment is running.
.IP "--unique_by_agent"
Orders in books are unique by agent : if the same agent places
another order, previous one is deleted. This is the default.
.IP "--no_unique_by_agent"
Agents may place multiple orders.
.IP "--csvdelimiter DELIMITER"
CSV files delimiter value.
.IP "--randomseed SEED"
Ramdom seed value.
.IP "--outputfilename OUTPUTFILE, -o OUTPUTFILE"
Set output filename. This overrides 
.I OUTPUTFILENAME
param in experiment config file.
.IP "--orderslogfile FILENAME, --ordersfile FILENAME"
Set orders log filename. This overrides 
.I ORDERSLOGFILE
param in experiment config file.
.SH COMMANDS
.IP "check"
Checks validity of 
.I experiment-config-file
by trying to instanciate all classes : world, engines, markets and agents.
.IP "run"
Run experiment from
.I experiment-config-file
.SH SEE ALSO
experiment-conf.yml(5)
.SH BUGS
Probably many.
.SH AUTHOR
Jean-Charles Bagneris <jcb at bagneris dot net>



fms.py
#!/usr/bin/env python
# -*- coding: utf8 -*-
"""
Financial Market Simulator
"""

__author__ = "Jean-Charles Bagneris <jcb@bagneris.net>"
__license__ = "GNU GPL v2"

import sys
import os.path
import optparse
import logging

from fms.utils import XmlParamsParser, YamlParamsParser

def set_parser():
    """
    Create command line arguments and options parser
    """
    optp = optparse.OptionParser(
        description='runs a Financial Market Simulator simulation',
        prog='%s' % sys.argv[0],
        usage="%prog [options] [command] simulationconffile",)
    optp.add_option('-v', '--verbose', action='store_true', 
        help="Set logging level to 'info'. Overrided by --loglevel.")
    optp.add_option('-L', '--loglevel', metavar='LEVEL', dest='loglevel',
        help="Set logging level to LEVEL: debug, info, warning, error, critical.")
    return optp

def set_logger(options):
    """
    Sets main logger instance.
    """
    levels = {'debug': logging.DEBUG,
              'info': logging.INFO,
              'warning': logging.WARNING,
              'error': logging.ERROR,
              'critical': logging.CRITICAL,}

    logger = logging.getLogger('fms')
    lhandler = logging.StreamHandler()
    formatter = logging.Formatter("%(levelname)s - %(name)s - %(message)s")
    lhandler.setFormatter(formatter)
    logger.addHandler(lhandler)
    logger.setLevel(logging.ERROR)
    if options.verbose:
        logger.setLevel(logging.INFO)
    if options.loglevel:
        logger.setLevel(levels[options.loglevel])
    return logger

def import_class(modulename, classname):
    """
    Try to import classname from modulename.
    """
    logger = logging.getLogger('fms')
    try:
        exec "from %s import %s as themodule " % \
                (modulename, classname.lower())
    except ImportError:
        logger.critical("Unknown %s class: %s" % (modulename, classname))
        sys.exit(2)
    return themodule

def set_world(params):
    """
    Import world class and instanciate world
    """
    logger = logging.getLogger('fms')
    worldmodule = import_class('fms.worlds', params['world']['classname'])
    world = getattr(worldmodule, params['world']['classname'])(params)
    logger.info("Created world %s" % world)
    return world

def set_agents(params):
    """
    Import agents class and instanciate agents
    """
    logger = logging.getLogger('fms')
    agentslist = []
    for a in params['agents']:
        agentmodule = import_class('fms.agents', a['classname'])
        for i in range(a['number']):
            agentslist.append(getattr(agentmodule, a['classname'])(a))
        logger.info("Created  %d instances of agent %s" % 
                (a['number'], agentslist[-1].__class__))
    return agentslist

def set_engines(params):
    """
    Import engines and markets classes and instanciate them
    """
    logger = logging.getLogger('fms')
    engineslist = []
    for (offset, e) in enumerate(params['engines']):
        marketmodule = import_class('fms.markets', e['market']['classname'])
        e['market']['instance'] = getattr(marketmodule, 
                e['market']['classname'])(params)
        enginemodule = import_class('fms.engines', e['classname'])
        e['instance'] = getattr(enginemodule, e['classname'])(params, offset)
        engineslist.append(e)
        logger.info("Created engine-market  %s - %s" % 
                (e['instance'], e['market']['instance']))
    return engineslist


def main():
    """
    Run experiment :
    - parse command line options and arguments
    - read simulation config file
    - call constructors for
      - world
      - engines and markets
      - agents
    - run [command]
    """
    parser = set_parser()
    options, arguments = parser.parse_args()

    logger = set_logger(options)

    try:
        command = arguments[0]
    except IndexError:
        parser.print_help()
        logger.critical("Missing command name.")
        return 2

    known_commands = ('run', 'check')
    if command not in known_commands:
        logger.critical("Unknown command: %s" % command)
        return 2

    try:
        simconffile = arguments[1]
    except IndexError:
        logger.critical("Missing simulation config file name.")
        return 2

    if os.path.splitext(simconffile)[-1] == '.xml':
        logger.debug("Calling XmlParamsParser on %s" % simconffile)
        params = XmlParamsParser(simconffile)
    else:
        logger.debug("Calling YamlParamsParser on %s" % simconffile)
        params = YamlParamsParser(simconffile)

    world = set_world(params)
    engineslist = set_engines(params)
    agentslist = set_agents(params)
            
    if command == 'run':
        logger.info("All is set, running simulation")
        for e in engineslist:
            logger.info("Running %s" % e['instance'])
            e['instance'].run(world, agentslist, e['market']['instance'])
        logger.info("Done.")

if __name__ == "__main__":
    sys.exit(main())
